File: backend/app/payment_helpers/northernMichigan/picturedRocks/tent/payFortSuperiorTent.py
from datetime import datetime
from playwright.sync_api import sync_playwright
import time


def pay_fortSuperiorTent(start_date, end_date, num_adults, num_kids, payment_info):
    start_timestamp = int(datetime.strptime(start_date, '%m/%d/%y').timestamp() * 1000)
    end_timestamp = int(datetime.strptime(end_date, '%m/%d/%y').timestamp() * 1000)

    response_data = {
        "base_price": 0,
        "tax": 0,
        "total": 0,
        "payment_successful": False
    }

    url = f"https://www.fortsuperiorcampground.com/campsites/rooms/%3FcheckIn%3D{start_timestamp}%26checkOut%3D{end_timestamp}%26adults%3D{num_adults}"

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context()
        page = context.new_page()

        try:
            page.goto(url, wait_until="domcontentloaded")

            iframe_element = page.wait_for_selector("iframe[title='Book a Room']", timeout=10000)
            if not iframe_element:
                raise Exception("Could not find iframe with title 'Book a Room'")

            iframe = iframe_element.content_frame()
            if not iframe:
                raise Exception("Could not get content frame from iframe")

            try:
                li_elements = iframe.wait_for_selector("li.room", state="attached", timeout=5000)
            except Exception as e:
                print("li_elements not found")
                return response_data

            li_elements = iframe.query_selector_all("li.room")

            for li in li_elements:
                site_name_element = li.query_selector("h3 a span.strans")

                if not site_name_element:
                    continue
                site_name = site_name_element.inner_text()
                if "Canvas Tent Barrack" not in site_name:
                    price_element = li.query_selector("div.price span.value")

                    if not price_element:
                        continue
                    price_text = price_element.inner_text()
                    book_now_button = li.query_selector("button.fancy-btn.s-button.button")
                    book_now_button.click()
                    time.sleep(1)

                    details_form = page.wait_for_selector("form[name='detailsForm']", timeout=10000)

                    first_name_input = details_form.query_selector("input#firstName")
                    first_name_input.fill(payment_info['first_name'])

                    last_name_input = details_form.query_selector("input#lastName")
                    last_name_input.fill(payment_info['last_name'])

                    email_input = details_form.query_selector("input#email")
                    email_input.fill(payment_info['email'])

                    phone_input = details_form.query_selector("input#phone")
                    phone_input.fill(payment_info['phone_number'])

                    country_select = details_form.query_selector("select#country")
                    country_select.select_option(value='string:us')

                    payment_div = page.wait_for_selector("div.payment", timeout=10000)
                    if payment_div:
                        # Get base price from first row of breakdown table
                        base_price_element = payment_div.query_selector("table.breakdown-table tr:first-child td:last-child")
                        if base_price_element:
                            base_price_text = base_price_element.inner_text().strip()
                            response_data["base_price"] = float(base_price_text.replace("$", ""))

                        # Get processing fee from second row
                        fee_element = payment_div.query_selector("table.breakdown-table tr:nth-child(2) td:last-child")
                        if fee_element:
                            fee_text = fee_element.inner_text().strip()
                            response_data["tax"] = float(fee_text.replace("$", ""))

                        # Get total from final row
                        total_element = payment_div.query_selector("tfoot tr td.total-value")
                        if total_element:
                            total_text = total_element.inner_text().strip()
                            response_data["total"] = float(total_text.replace("$", ""))

                    continue_button = details_form.query_selector("button.pay-now")
                    continue_button.click()
                    time.sleep(1)

                    form_details = page.wait_for_selector("div#form-details", timeout=10000)

                    # Switch to the iframe with title='Credit / Debit Card'
                    iframe = page.wait_for_selector("iframe[title='Credit / Debit Card']", timeout=10000)
                    iframe_content = iframe.content_frame()

                    # Fill out the card details within the iframe
                    iframe_content.fill("input#cardNumber", payment_info['card_number'])
                    iframe_content.fill("input#cardExpiration", payment_info['expiry_date'])
                    iframe_content.fill("input#cardCvv", payment_info['cvc'])
                    iframe_content.fill("input#cardHolderName", payment_info['cardholder_name'])

                    # Continue with submitting the payment if necessary
                    complete_checkout_button = form_details.query_selector("button[data-hook='cashier-payments-method-pay-button']")
                    # complete_checkout_button.click() #uncomment to pay
                    response_data['payment_successful'] = True
                    return response_data

            
            return response_data

        except Exception as e:
            print(f"An error occurred: {e}")
            return response_data

        finally:
            browser.close()


def main():
    payment_info = {
        "first_name": "Lebron",
        "last_name": "James",
        "email": "lebronjames@gmail.com",
        "phone_number": "3134321234",
        "street_address": "1234 Rocky Rd",
        "city": "San Francisco",
        "state": "CA",
        "zip_code": "45445",
        "country": "USA",
        "cardholder_name": "Lebron James",
        "card_number": "2342943844322224",
        "card_type": "Visa",
        "expiry_date": "01/30",
        "cvc": "1234"
    }
    fortSuperiorData = pay_fortSuperiorTent('06/08/25', '06/10/25', 2, 1, payment_info)
    print(fortSuperiorData)

if __name__ == '__main__':
    main()

File: backend/app/scrapers-local-testing/pay_fortSuperiorTent.js
const puppeteer = require('puppeteer');

async function payFortSuperiorTent(startDate, endDate, numAdults, numKids, paymentInfo) {
  const responseData = {
    base_price: 0,
    tax: 0,
    total: 0,
    payment_successful: false,
  };

  // Format dates to timestamps like in Python version
  const startTimestamp = new Date(startDate).getTime();
  const endTimestamp = new Date(endDate).getTime();
  const num_travelers = numAdults + numKids;
  
  const url = `https://www.fortsuperiorcampground.com/campsites/rooms/%3FcheckIn%3D${startTimestamp}%26checkOut%3D${endTimestamp}%26adults%3D${num_travelers}`;

  try {
    const browser = await puppeteer.launch({
      headless: true,
      args: ["--no-sandbox", "--disable-setuid-sandbox"]
    });

    const page = await browser.newPage();
    await page.setViewport({ width: 1920, height: 1080 });
    console.log('Browser launched with viewport 1920x1080');

    await page.goto(url, { waitUntil: 'networkidle2' });
    console.log(`Navigated to URL: ${url}`);
    await new Promise(resolve => setTimeout(resolve, 1000));
    
    // Wait for and get iframe
    const frameHandle = await page.waitForSelector("iframe[title='Book a Room']");
    const frame = await frameHandle.contentFrame();
    console.log('Found and switched to booking iframe');
    
    // Find room listings
    const liElements = await frame.$$("li.room");
    console.log(`Found ${liElements.length} room listings`);
    
    for (const li of liElements) {
      const siteNameElement = await li.$("h3 a span.strans");
      if (!siteNameElement) {
        console.log('Skipping listing - no site name element found');
        continue;
      }
      
      const siteName = await siteNameElement.evaluate(el => el.innerText);
      console.log(`Processing room: ${siteName}`);
      if (!siteName.includes("Canvas Tent Barrack")) {
        const priceElement = await li.$("div.price span.value");
        if (!priceElement) continue;
        
        const bookButton = await li.$("button.fancy-btn.s-button.button");
        await bookButton.evaluate(el => el.scrollIntoView());
        await bookButton.evaluate(b => b.click());
        console.log('Clicked book button');
        await new Promise(resolve => setTimeout(resolve, 1000));

        // Fill out details form
        const detailsForm = await page.waitForSelector("form[name='detailsForm']");
        console.log('Found details form, filling out personal information');
        await page.type("input#firstName", paymentInfo.first_name);
        await page.type("input#lastName", paymentInfo.last_name);
        await page.type("input#email", paymentInfo.email);
        await page.type("input#phone", paymentInfo.phone_number);
        
        await page.select("select#country", 'string:us');

        // Get pricing information
        const paymentDiv = await page.waitForSelector("div.payment");
        const basePriceEl = await paymentDiv.$("table.breakdown-table tr:first-child td:last-child");
        if (basePriceEl) {
          const basePrice = await basePriceEl.evaluate(el => parseFloat(el.innerText.replace('$', '')));
          responseData.base_price = basePrice;
          console.log(`Base price: $${basePrice}`);
        }

        const feeEl = await paymentDiv.$("table.breakdown-table tr:nth-child(2) td:last-child");
        if (feeEl) {
          const fee = await feeEl.evaluate(el => parseFloat(el.innerText.replace('$', '')));
          responseData.tax = fee;
        }

        const totalEl = await paymentDiv.$("tfoot tr td.total-value");
        if (totalEl) {
          const total = await totalEl.evaluate(el => parseFloat(el.innerText.replace('$', '')));
          responseData.total = total;
        }

        // Continue to payment
        const continueButton = await detailsForm.$("button.pay-now");
        await continueButton.click();
        await new Promise(resolve => setTimeout(resolve, 1000));

        // Handle credit card iframe
        const cardFrame = await page.waitForSelector("iframe[title='Credit / Debit Card']");
        const cardFrameContent = await cardFrame.contentFrame();
        console.log('Found and switched to credit card iframe');
        
        await new Promise(resolve => setTimeout(resolve, 5000)); // Wait for card number field to be ready
        await cardFrameContent.waitForSelector("input#cardNumber", { timeout: 10000 });
        await cardFrameContent.type("input#cardNumber", paymentInfo.card_number);
        await new Promise(resolve => setTimeout(resolve, 1000));
        
        await cardFrameContent.type("input#cardExpiration", paymentInfo.expiry_date);
        await new Promise(resolve => setTimeout(resolve, 1000));
        
        await cardFrameContent.type("input#cardCvv", paymentInfo.cvc);
        await new Promise(resolve => setTimeout(resolve, 1000));
        
        // Clear cardholder name field before typing
        await cardFrameContent.evaluate(() => document.getElementById('cardHolderName').value = '');
        await cardFrameContent.type("input#cardHolderName", paymentInfo.cardholder_name);
        await new Promise(resolve => setTimeout(resolve, 1000));
        
        console.log('Filled out payment information');

        const checkoutButton = await page.$("button[data-hook='cashier-payments-method-pay-button']");
        // await checkoutButton.click(); // uncomment to actually pay
        
        responseData.payment_successful = true;
        await browser.close();
        return responseData;
      }
    }

    await browser.close();
    return responseData;

  } catch (error) {
    console.error(`Error during payment: ${error.message}`);
    return responseData;
  }
}

if (require.main === module) {
    (async () => {
        const paymentInfo = {
            first_name: "Lebron",
            last_name: "James",
            email: "lebronjames@gmail.com",
            phone_number: "3134321234",
            street_address: "1234 Rocky Rd",
            city: "San Francisco",
            state: "CA",
            zip_code: "45445",
            country: "USA",
            cardholder_name: "Lebron James",
            card_number: "2342943844322224",
            card_type: "Visa",
            expiry_date: "01/30",
            cvc: "1234"
        }
        const result = await payFortSuperiorTent("06/08/25", "06/10/25", 2, 1, paymentInfo);
        console.log("Scrape result:", result);
    })();
}

module.exports = { payFortSuperiorTent };